src/main/java/com/salesforce/dynamodbv2/mt/util/RecordsCache.java

```diff
@@ -0,0 +1,276 @@
+package com.salesforce.dynamodbv2.mt.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.collect.Iterables.getLast;
+
+import com.amazonaws.services.dynamodbv2.model.Record;
+import com.amazonaws.services.dynamodbv2.model.StreamRecord;
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import javax.annotation.Nonnull;
+
+class RecordsCache {
+
+    /**
+     * A cached segment of a stream shard. Begins with {@link #start} (inclusive) and ends with {@link #end} exclusive.
+     * The {@link #records} collection is sorted by record sequence number and may be empty. All sequence numbers in the
+     * collection fall into the segment sequence number range.
+     */
+    private static final class Segment {
+
+        @Nonnull
+        private final SequenceNumber start;
+        @Nonnull
+        private final SequenceNumber end;
+        @Nonnull
+        private final List<Record> records;
+        @Nonnull
+        private final long byteSize;
+
+        Segment(SequenceNumber start, List<Record> records) {
+            this(start, SequenceNumber.fromRecord(getLast(records)).next(), records);
+        }
+
+        Segment(SequenceNumber start, SequenceNumber end, List<Record> records) {
+            assert start.compareTo(end) <= 0;
+            this.start = checkNotNull(start);
+            this.end = checkNotNull(end);
+            this.records = Collections.unmodifiableList(checkNotNull(records));
+            this.byteSize = records.stream().map(Record::getDynamodb).mapToLong(StreamRecord::getSizeBytes).sum();
+        }
+
+        @Nonnull
+        public SequenceNumber getStart() {
+            return start;
+        }
+
+        @Nonnull
+        public SequenceNumber getEnd() {
+            return end;
+        }
+
+        @Nonnull
+        public List<Record> getRecords() {
+            return records;
+        }
+
+        List<Record> getRecords(SequenceNumber from) {
+            assert start.compareTo(from) <= 0 && end.compareTo(from) >= 0;
+
+            if (records.isEmpty()) {
+                return records;
+            }
+
+            if (start.equals(from)) {
+                return records;
+            }
+
+            return records.subList(getIndex(from), records.size());
+        }
+
+        private int getIndex(SequenceNumber sequenceNumber) {
```

Comment:
This is another behavior change (though non-functional) compared to master: instead of iterating over all records to find the index of a sequence number in a list of records, this impl uses binary search + lazy parsing of string sequence numbers into BigIntegers to allow comparison. I am hoping that will reduce the CPU load on our streams dynos.